transport/src/main/java/io/el/channel/DefaultChannelPipeline.java

```diff
@@ -0,0 +1,278 @@
+package io.el.channel;
+
+import io.el.concurrent.EventLoop;
+import io.el.internal.ObjectUtil;
+import java.net.SocketAddress;
+import java.util.UUID;
+
+/**
+ * {@link DefaultChannelPipeline} manages a channel handler list. When an event occurs, it calls the
+ * handlers it manages.
+ */
+public class DefaultChannelPipeline implements ChannelPipeline {
+
+  private final Channel channel;
+  private final HeadContext headContext;
+  private final TailContext tailContext;
+  private static final String HEAD_NAME = "HeaderContext";
+  private static final String TAIL_NAME = "TailContext";
+
+  public DefaultChannelPipeline(Channel channel) {
+    this.channel = channel;
+    this.tailContext = new TailContext();
+    this.headContext = new HeadContext();
+
+    this.tailContext.prev = this.headContext;
+    this.headContext.next = this.tailContext;
+  }
+
+  @Override
+  public Channel channel() {
+    return this.channel;
+  }
+
+  @Override
+  public ChannelPipeline addLast(ChannelHandler... handlers) {
+    for (ChannelHandler handler : handlers) {
+      this.addLast(handler);
+    }
+    return this;
+  }
+
+  private ChannelPipeline addLast(ChannelHandler handler) {
+    final AbstractChannelHandlerContext prev = this.tailContext.prev;
+    // TODO: generate name with the {@link ChannelHandler.getClass()}
+    String name = UUID.randomUUID().toString();
+    EventLoop eventLoop = null;
+    final AbstractChannelHandlerContext newHandlerContext = new DefaultHandlerContext__(name, this, eventLoop, handler);
+
+
+    this.tailContext.prev = newHandlerContext;
+    newHandlerContext.next = this.tailContext;
+
+    prev.next = newHandlerContext;
+    newHandlerContext.prev = prev;
+    return this;
+  }
+
+  @Override
+  public ChannelPipeline remove(ChannelHandler handler) {
+    final AbstractChannelHandlerContext context = this.context(handler);
+    remove(context);
+    return this;
+  }
+
+  private void remove(AbstractChannelHandlerContext context) {
+    assert context != this.headContext && context != this.tailContext;
+
+    // We need synchronize to call handler removed event atomically.
+    synchronized (this) {
+      atomicRemoveFromHeandlerList(context);
+
+      // TODO: call handler removed
```

Comment:
In the first spec, we don't have handler removed event so please remove it with synchronized block